packages/material-ui-styles/src/makeStyles.js

```diff
@@ -11,19 +14,181 @@ const noopTheme = {};
+
+    if (stylesOptions.sheetsCache) {
+      staticSheet = multiKeyStore.get(stylesOptions.sheetsCache, stylesCreator, theme);
+    }
+
+    const styles = stylesCreator.create(theme, name);
+
+    if (!staticSheet) {
+      staticSheet = stylesOptions.jss.createStyleSheet(styles, {
+        link: false,
+        ...options,
+      });
+
+      staticSheet.attach();
+
+      if (stylesOptions.sheetsCache) {
+        multiKeyStore.set(stylesOptions.sheetsCache, stylesCreator, theme, staticSheet);
+      }
+    }
+
+    if (sheetsRegistry) {
+      sheetsRegistry.add(staticSheet);
+    }
+
+    sheetManager.dynamicStyles = getDynamicStyles(styles);
+    sheetManager.staticSheet = staticSheet;
+  }
+
+  if (sheetManager.dynamicStyles) {
+    const dynamicSheet = stylesOptions.jss.createStyleSheet(sheetManager.dynamicStyles, {
+      link: true,
+      ...options,
+    });
+
+    warning(props, 'Material-UI: properties missing.');
+
+    dynamicSheet.update(props).attach();
+
+    state.dynamicSheet = dynamicSheet;
+
+    if (sheetsRegistry) {
+      sheetsRegistry.add(dynamicSheet);
+    }
+
+    state.classes = mergeClasses({
+      baseClasses: sheetManager.staticSheet.classes,
+      newClasses: dynamicSheet.classes,
+    });
+  } else {
+    state.classes = sheetManager.staticSheet.classes;
+  }
+
+  sheetManager.refs += 1;
+}
+
+function update({ state, props }) {
+  if (state.dynamicSheet) {
+    state.dynamicSheet.update(props);
+  }
+}
+
+function detach({ state, theme, stylesOptions, stylesCreator }) {
+  if (stylesOptions.disableGeneration) {
+    return;
+  }
+
+  const sheetManager = multiKeyStore.get(stylesOptions.sheetsManager, stylesCreator, theme);
+  sheetManager.refs -= 1;
+  const sheetsRegistry = stylesOptions.sheetsRegistry;
+
+  if (sheetManager.refs === 0) {
+    multiKeyStore.delete(stylesOptions.sheetsManager, stylesCreator, theme);
+    stylesOptions.jss.removeStyleSheet(sheetManager.staticSheet);
+
+    if (sheetsRegistry) {
+      sheetsRegistry.remove(sheetManager.staticSheet);
+    }
+  }
+
+  if (state.dynamicSheet) {
+    stylesOptions.jss.removeStyleSheet(state.dynamicSheet);
+
+    if (sheetsRegistry) {
+      sheetsRegistry.remove(state.dynamicSheet);
+    }
+  }
+}
+
 function makeStyles(stylesOrCreator, options = {}) {
-  const { withTheme = false, name, defaultTheme: defaultThemeOption, ...stylesOptions2 } = options;
+  const {
+    // An explicit value provided by the developers.
+    name,
+    // Help with debuggability.
+    meta: metaOption,
+    defaultTheme: defaultThemeOption,
+    Component,
+    ...stylesOptions2
+  } = options;
   const stylesCreator = getStylesCreator(stylesOrCreator);
-  const listenToTheme = stylesCreator.themingEnabled || typeof name === 'string' || withTheme;
+  const listenToTheme = stylesCreator.themingEnabled || typeof name === 'string';
   const defaultTheme = defaultThemeOption || noopTheme;
 
-  const meta = name || 'Hook';
+  const meta = name || metaOption || 'Hook';
```

Comment:
What is the use case for this?

Reply:
Commenting the diff would make two tests fail:
```
24 passing (2s)
  2 failing

  1) withStyles
       integration
         should run lifecycles with no theme:

      AssertionError: expected { root: 'Hook-root-vy1bts' } to deeply equal { root: 'Empty-root-vy1bts' }
      + expected - actual

       {
      -  "root": "Hook-root-vy1bts"
      +  "root": "Empty-root-vy1bts"
       }

      at Context.deepEqual (packages/material-ui-styles/src/withStyles.test.js:80:14)

  2) withStyles
       classname quality
         should use the displayName:

      AssertionError: expected 'Hook' to equal 'a'
      + expected - actual

      -Hook
      +a

      at Context.strictEqual (packages/material-ui-styles/src/withStyles.test.js:219:14)
```

withStyles uses the display name to provide a meta option.

Reply:
Shouldn't it just override the `name` then? It seems like the `name` and `meta` option are doing the same.

Reply:
The name value is human picked. The meta is inferred. The name can have a strong implication on the generated class names, the meta is only used for improving the debugging experience.

Reply:
Can we change the name of `meta` then? It's a pretty generic term. What speaks against using `inferredName`?

Reply:
The meta comes from JSS: https://cssinjs.org/jss-api?v=v10.0.0-alpha.10#create-style-sheet. We can replace it with classNamePrefix in this case?